src/stores/workspace/WorkspaceViewStore.ts

```diff
@@ -59,13 +62,165 @@ export default class WorkspaceViewStore {
 	};
 
 	@action
-	public resetToDefaulLayout = () => {
-		this.panelsLayout = [50, 50];
+	public setCollapsedPanels = (collapsedPanels: number[]) => {
+		this.collapsedPanels = collapsedPanels;
 	};
 
 	@action
 	public collapsePanel = (index: number) => {
-		this.panelsLayout = index === 0 ? [0, 100] : [100, 0];
+		if (this.collapsedPanels.length === 0) {
```

Comment:
do we really need `collapsedPanels` state? cant we just tell from panels width if its collapsed or not?
also algorithm looks rather complicated and long, i'm sure there must be a better way to do it

Reply:
```
		const layouts = [[100], [50, 50], [35, 35, 30], [35, 25, 20, 20]];
		const notCollapsedPanels = this.panelsLayout
			.map((w, i) => (w > 10 ? i : null))
			.filter(w => w !== null);
		const isTargetPanelCollapsed = this.panelsLayout[index] < 10;
		const panelCount = notCollapsedPanels.includes(index)
			? notCollapsedPanels.length - 1
			: notCollapsedPanels.length + 1;

		if (panelCount <= 0) return;
		const widths = layouts[panelCount - 1];

		const updatedLayout = [];
		for (let i = 0; i < this.panelsLayout.length; i++) {
			let width: number | undefined = 0;
			if (i === index) {
				width = isTargetPanelCollapsed ? widths.shift() : 0;
			} else if (this.panelsLayout[i] > 10) {
				width = widths.shift();
			}
			updatedLayout.push(width || 0);
		}

		this.panelsLayout = updatedLayout as any;
```